File: frontend/src/types/content.ts
/**
 * KnowledgeTree Content Workbench Types
 * TypeScript interfaces for Content Editor, Versioning, AI Operations, and Templates
 */

// Content Status Enum
export const ContentStatus = {
  DRAFT: 'draft',
  REVIEW: 'review',
  PUBLISHED: 'published',
} as const;

export type ContentStatus = typeof ContentStatus[keyof typeof ContentStatus];

// Tone Types for AI Rewriting
export const ToneType = {
  PROFESSIONAL: 'professional',
  CASUAL: 'casual',
  TECHNICAL: 'technical',
  FRIENDLY: 'friendly',
  FORMAL: 'formal',
  CONVERSATIONAL: 'conversational',
} as const;

export type ToneType = typeof ToneType[keyof typeof ToneType];

// Reading Levels for Simplification
export const ReadingLevel = {
  BASIC: 'basic',
  INTERMEDIATE: 'intermediate',
  ADVANCED: 'advanced',
} as const;

export type ReadingLevel = typeof ReadingLevel[keyof typeof ReadingLevel];

// Quote Types
export const QuoteType = {
  FACT: 'fact',
  OPINION: 'opinion',
  DEFINITION: 'definition',
  STATISTIC: 'statistic',
  EXAMPLE: 'example',
} as const;

export type QuoteType = typeof QuoteType[keyof typeof QuoteType];

// Template Types
export const TemplateType = {
  HOW_TO: 'how_to',
  FAQ: 'faq',
  TUTORIAL: 'tutorial',
  ARTICLE: 'article',
  REFERENCE: 'reference',
} as const;

export type TemplateType = typeof TemplateType[keyof typeof TemplateType];

// Extended Category with Content Workbench fields
export interface CategoryWithContent {
  id: number;
  name: string;
  description: string | null;
  color: string;
  icon: string;
  depth: number;
  order: number;
  parent_id: number | null;
  project_id: number;
  merged_content: string | null;

  // Content Workbench fields
  draft_content: string | null;
  published_content: string | null;
  content_status: ContentStatus;
  published_at: string | null;
  reviewed_by: number | null;
  reviewed_at: string | null;

  created_at: string;
  updated_at: string;
}

// Content Version
export interface ContentVersion {
  id: number;
  category_id: number;
  version_number: number;
  content: string;
  created_by: number | null;
  created_at: string;
  change_summary: string | null;
}

// Extracted Quote
export interface ExtractedQuote {
  id: number;
  category_id: number;
  quote_text: string;
  context_before: string | null;
  context_after: string | null;
  quote_type: QuoteType | null;
  created_at: string;
}

// Content Template
export interface ContentTemplate {
  id: number;
  name: string;
  description: string | null;
  template_type: TemplateType;
  structure: {
    sections: TemplateSection[];
  };
  is_public: boolean;
  created_by: number | null;
  created_at: string;
}

export interface TemplateSection {
  title: string;
  prompt: string;
  order: number;
}

// Request Types

export interface SaveDraftRequest {
  draft_content: string;
  change_summary?: string;
  auto_version?: boolean;
}

export interface PublishContentRequest {
  create_version?: boolean;
}

export interface RestoreVersionRequest {
  version_number: number;
  create_new_version?: boolean;
}

export interface CompareVersionsRequest {
  version_a: number;
  version_b: number;
}

export interface SummarizeRequest {
  content: string;
  max_length?: number;
  focus?: string;
}

export interface ExpandRequest {
  content: string;
  target_length?: number;
  add_details?: string;
}

export interface SimplifyRequest {
  content: string;
  reading_level?: ReadingLevel;
}

export interface RewriteToneRequest {
  content: string;
  tone: ToneType;
  preserve_facts?: boolean;
}

export interface ExtractQuotesRequest {
  content: string;
  max_quotes?: number;
  quote_types?: QuoteType[];
}

export interface GenerateOutlineRequest {
  topic: string;
  depth?: number;
  style?: TemplateType;
}

export interface ContentTemplateCreateRequest {
  name: string;
  description?: string;
  template_type: TemplateType;
  sections: TemplateSectionCreate[];
  is_public?: boolean;
}

export interface TemplateSectionCreate {
  title: string;
  prompt: string;
  order: number;
}

// Response Types

export interface VersionComparisonResponse {
  version_a: {
    number: number;
    content: string;
    created_at: string;
    created_by: number | null;
    change_summary: string | null;
  };
  version_b: {
    number: number;
    content: string;
    created_at: string;
    created_by: number | null;
    change_summary: string | null;
  };
}

export interface AIOperationResponse {
  result: string;
  operation: string;
  tokens_used: number | null;
}

File: backend/schemas/content.py
"""
Content Workbench Schemas
Pydantic models for content editing, versioning, and AI operations
"""

from pydantic import BaseModel, Field
from typing import Optional, List, Dict, Any
from datetime import datetime
from enum import Enum


# Enums

class ContentStatus(str, Enum):
    """Content workflow status"""
    DRAFT = "draft"
    REVIEW = "review"
    PUBLISHED = "published"


class ToneType(str, Enum):
    """Available tone types for rewriting"""
    PROFESSIONAL = "professional"
    CASUAL = "casual"
    TECHNICAL = "technical"
    FRIENDLY = "friendly"
    FORMAL = "formal"
    CONVERSATIONAL = "conversational"


class ReadingLevel(str, Enum):
    """Target reading levels for simplification"""
    BASIC = "basic"
    GENERAL = "general"
    ADVANCED = "advanced"


class QuoteType(str, Enum):
    """Types of extracted quotes"""
    FACT = "fact"
    OPINION = "opinion"
    DEFINITION = "definition"
    STATISTIC = "statistic"
    EXAMPLE = "example"


class TemplateType(str, Enum):
    """Content template categories"""
    HOW_TO = "how_to"
    FAQ = "faq"
    TUTORIAL = "tutorial"
    ARTICLE = "article"
    REFERENCE = "reference"


# Content Version Schemas

class ContentVersionBase(BaseModel):
    """Base content version schema"""
    content: str = Field(..., min_length=1)
    change_summary: Optional[str] = Field(None, max_length=500)


class ContentVersionResponse(BaseModel):
    """Content version response"""
    id: int
    category_id: int
    version_number: int
    content: str
    created_by: Optional[int]
    created_at: datetime
    change_summary: Optional[str]

    model_config = {
        "from_attributes": True,
        "json_schema_extra": {
            "examples": [
                {
                    "id": 1,
                    "category_id": 42,
                    "version_number": 5,
                    "content": "# Updated Content\n\nThis is version 5...",
                    "created_by": 1,
                    "created_at": "2026-02-02T12:00:00Z",
                    "change_summary": "Added new section on security"
                }
            ]
        }
    }


# Extracted Quote Schemas

class ExtractedQuoteResponse(BaseModel):
    """Extracted quote response"""
    id: int
    category_id: int
    quote_text: str
    context_before: Optional[str]
    context_after: Optional[str]
    quote_type: Optional[str]
    created_at: datetime

    model_config = {
        "from_attributes": True,
        "json_schema_extra": {
            "examples": [
                {
                    "id": 1,
                    "category_id": 42,
                    "quote_text": "Security is not a feature, it's a foundation.",
                    "context_before": "When building applications, many developers...",
                    "context_after": "This means implementing security from day one...",
                    "quote_type": "opinion",
                    "created_at": "2026-02-02T12:00:00Z"
                }
            ]
        }
    }


# Content Template Schemas

class TemplateSectionCreate(BaseModel):
    """Template section structure"""
    title: str = Field(..., min_length=1, max_length=200)
    prompt: str = Field(..., min_length=1, max_length=1000)
    order: int = Field(..., ge=0)


class ContentTemplateCreate(BaseModel):
    """Schema for creating a content template"""
    name: str = Field(..., min_length=1, max_length=200)
    description: Optional[str] = None
    template_type: TemplateType
    sections: List[TemplateSectionCreate] = Field(..., min_items=1)
    is_public: bool = Field(default=True)

    model_config = {
        "json_schema_extra": {
            "examples": [
                {
                    "name": "How-To Guide Template",
                    "description": "Standard template for creating how-to guides",
                    "template_type": "how_to",
                    "sections": [
                        {
                            "title": "Introduction",
                            "prompt": "Explain what this guide covers and who it's for",
                            "order": 1
                        },
                        {
                            "title": "Prerequisites",
                            "prompt": "List required knowledge, tools, and setup",
                            "order": 2
                        },
                        {
                            "title": "Step-by-Step Instructions",
                            "prompt": "Provide detailed, numbered steps",
                            "order": 3
                        },
                        {
                            "title": "Troubleshooting",
                            "prompt": "Address common issues and solutions",
                            "order": 4
                        }
                    ],
                    "is_public": True
                }
            ]
        }
    }


class ContentTemplateResponse(BaseModel):
    """Content template response"""
    id: int
    name: str
    description: Optional[str]
    template_type: str
    structure: Dict[str, Any]  # JSONB field
    is_public: bool
    created_by: Optional[int]
    created_at: datetime

    model_config = {
        "from_attributes": True,
        "json_schema_extra": {
            "examples": [
                {
                    "id": 1,
                    "name": "How-To Guide Template",
                    "description": "Standard template for creating how-to guides",
                    "template_type": "how_to",
                    "structure": {
                        "sections": [
                            {
                                "title": "Introduction",
                                "prompt": "Explain what this guide covers and who it's for",
                                "order": 1
                            }
                        ]
                    },
                    "is_public": True,
                    "created_by": 1,
                    "created_at": "2026-02-02T12:00:00Z"
                }
            ]
        }
    }


# AI Rewriting Request Schemas

class SummarizeRequest(BaseModel):
    """Request for content summarization"""
    content: str = Field(..., min_length=1)
    max_length: Optional[int] = Field(None, ge=10, le=1000)
    focus: Optional[str] = Field(None, max_length=200)

    model_config = {
        "json_schema_extra": {
            "examples": [
                {
                    "content": "Long article text here...",
                    "max_length": 150,
                    "focus": "technical implementation"
                }
            ]
        }
    }


class ExpandRequest(BaseModel):
    """Request for content expansion"""
    content: str = Field(..., min_length=1)
    target_length: Optional[int] = Field(None, ge=100, le=5000)
    add_details: Optional[str] = Field(None, max_length=200)

    model_config = {
        "json_schema_extra": {
            "examples": [
                {
                    "content": "Brief summary to expand...",
                    "target_length": 500,
                    "add_details": "examples and code snippets"
                }
            ]
        }
    }


class SimplifyRequest(BaseModel):
    """Request for content simplification"""
    content: str = Field(..., min_length=1)
    reading_level: ReadingLevel = Field(default=ReadingLevel.GENERAL)

    model_config = {
        "json_schema_extra": {
            "examples": [
                {
                    "content": "Complex technical text here...",
                    "reading_level": "basic"
                }
            ]
        }
    }


class RewriteToneRequest(BaseModel):
    """Request for tone rewriting"""
    content: str = Field(..., min_length=1)
    tone: ToneType
    preserve_facts: bool = Field(default=True)

    model_config = {
        "json_schema_extra": {
            "examples": [
                {
                    "content": "Original text here...",
                    "tone": "professional",
                    "preserve_facts": True
                }
            ]
        }
    }


class ExtractQuotesRequest(BaseModel):
    """Request for quote extraction"""
    content: str = Field(..., min_length=1)
    max_quotes: int = Field(default=10, ge=1, le=50)
    quote_types: Optional[List[QuoteType]] = None

    model_config = {
        "json_schema_extra": {
            "examples": [
                {
                    "content": "Document content to extract quotes from...",
                    "max_quotes": 10,
                    "quote_types": ["fact", "statistic"]
                }
            ]
        }
    }


class GenerateOutlineRequest(BaseModel):
    """Request for outline generation"""
    topic: str = Field(..., min_length=1, max_length=500)
    depth: int = Field(default=2, ge=1, le=3)
    style: TemplateType = Field(default=TemplateType.ARTICLE)

    model_config = {
        "json_schema_extra": {
            "examples": [
                {
                    "topic": "Introduction to Microservices Architecture",
                    "depth": 2,
                    "style": "tutorial"
                }
            ]
        }
    }


# AI Operation Response Schemas

class AIOperationResponse(BaseModel):
    """Generic AI operation response"""
    result: str
    operation: str
    tokens_used: Optional[int] = None

    model_config = {
        "json_schema_extra": {
            "examples": [
                {
                    "result": "Transformed content here...",
                    "operation": "summarize",
                    "tokens_used": 450
                }
            ]
        }
    }


# Content Editor Request Schemas

class SaveDraftRequest(BaseModel):
    """Request to save draft content"""
    draft_content: str = Field(..., min_length=1)
    change_summary: Optional[str] = Field(None, max_length=500)
    auto_version: bool = Field(default=True)

    model_config = {
        "json_schema_extra": {
            "examples": [
                {
                    "draft_content": "# Updated Content\n\nNew section added...",
                    "change_summary": "Added troubleshooting section",
                    "auto_version": True
                }
            ]
        }
    }


class PublishContentRequest(BaseModel):
    """Request to publish content"""
    create_version: bool = Field(default=True)


class RestoreVersionRequest(BaseModel):
    """Request to restore a version"""
    version_number: int = Field(..., ge=1)
    create_new_version: bool = Field(default=True)


class CompareVersionsRequest(BaseModel):
    """Request to compare two versions"""
    version_a: int = Field(..., ge=1)
    version_b: int = Field(..., ge=1)


class VersionComparisonResponse(BaseModel):
    """Version comparison response"""
    version_a: Dict[str, Any]
    version_b: Dict[str, Any]

    model_config = {
        "json_schema_extra": {
            "examples": [
                {
                    "version_a": {
                        "number": 3,
                        "content": "Version 3 content...",
                        "created_at": "2026-02-01T10:00:00Z",
                        "created_by": 1,
                        "change_summary": "Initial version"
                    },
                    "version_b": {
                        "number": 5,
                        "content": "Version 5 content...",
                        "created_at": "2026-02-02T12:00:00Z",
                        "created_by": 1,
                        "change_summary": "Added security section"
                    }
                }
            ]
        }
    }
